### PR TITLE
common/math, core/vm: implement fast EXP.

### PR DESCRIPTION
**Pull request by obscuren**
Courtesy @chfast & @karalabe

### CODE REVIEW

File: common/math/exp.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const (
+	// Perverted constants straight out of big.Int
+	// https://golang.org/src/math/big/arith.go
+	_m    = ^big.Word(0)
+	_logS = _m>>8&1 + _m>>16&1 + _m>>32&1
+	_S    = 1 << _logS
+	_W    = _S << 3 // word size in bits
```
Comment (fjl) [inline common/math/exp.go]: Please use:

```go
const wordSize = 32 << big.Word(^big.Word(0)>>63)
```

as requested on the other PR.

Comment (chfast) [inline common/math/exp.go]: You can also use `unsafe.SizeOf`.

Comment (fjl) [inline common/math/exp.go]: Yes, but I'd prefer a const calculation package math. There is no downside and importing unsafe makes the code less portable (e.g. makes compiling to JS impossible).

**Comment by karalabe** [comment]
Also take care as compilation fails on CIs currently with some type casting error.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ed3264e3db0e46e13cfcf1e65f79e1592a96fe9e
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ed3264e3db0e46e13cfcf1e65f79e1592a96fe9e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 0700064c598b621985b7c70a68496608745eee29
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ed3264e3db0e46e13cfcf1e65f79e1592a96fe9e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 0700064c598b621985b7c70a68496608745eee29
  * Commits must be prefixed with the package(s) they modify

 * Commit: 88f095aae8c1d23275b7086f0cd49851fcdc539b
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ed3264e3db0e46e13cfcf1e65f79e1592a96fe9e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 0700064c598b621985b7c70a68496608745eee29
  * Commits must be prefixed with the package(s) they modify

 * Commit: 88f095aae8c1d23275b7086f0cd49851fcdc539b
  * Commits must be prefixed with the package(s) they modify

 * Commit: 147bc0073378bf294d4f1b92e8ff5aa5e427d1e9
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by karalabe** [comment]
@fjl Please double check, I changed the formula to avoid type size issues from `go vet`. If it look good please squash & merge.

**Comment by karalabe** [comment]
Ok, someone deleted all the fix commits from this branch. Fess up :P I pushed the last fix we did back now.

File: common/math/exp.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package math
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// wordSize is the size number of bytes in a big.Int Word.
```
Comment (fjl) [inline common/math/exp.go]: It's actually the number of bits.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ab18edbedc12f141a803af3a241c67d24cc9e518
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Pull request merged**